<!DOCTYPE html PUBLIC ""
    "">
<html><head><meta charset="UTF-8" /><title>aerial.bio.utils.files documentation</title><link rel="stylesheet" type="text/css" href="css/default.css" /><link rel="stylesheet" type="text/css" href="css/highlight.css" /><script type="text/javascript" src="js/highlight.min.js"></script><script type="text/javascript" src="js/jquery.min.js"></script><script type="text/javascript" src="js/page_effects.js"></script><script>hljs.initHighlightingOnLoad();</script></head><body><div id="header"><h2>Generated by <a href="https://github.com/weavejester/codox">Codox</a></h2><h1><a href="index.html"><span class="project-title"><span class="project-name">Aerial.bio.utils</span> <span class="project-version">1.0.0</span></span></a></h1></div><div class="sidebar primary"><h3 class="no-link"><span class="inner">Project</span></h3><ul class="index-link"><li class="depth-1 "><a href="index.html"><div class="inner">Index</div></a></li></ul><h3 class="no-link"><span class="inner">Topics</span></h3><ul><li class="depth-1 "><a href="intro.html"><div class="inner"><span>Introduction to bio.utils</span></div></a></li></ul><h3 class="no-link"><span class="inner">Namespaces</span></h3><ul><li class="depth-1"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>aerial</span></div></div></li><li class="depth-2"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>bio</span></div></div></li><li class="depth-3"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>utils</span></div></div></li><li class="depth-4 branch"><a href="aerial.bio.utils.aligners.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>aligners</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.extools.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>extools</span></div></a></li><li class="depth-4 branch current"><a href="aerial.bio.utils.files.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>files</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.filters.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>filters</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.infoth.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>infoth</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.params.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>params</span></div></a></li><li class="depth-4"><a href="aerial.bio.utils.seqxform.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>seqxform</span></div></a></li></ul></div><div class="sidebar secondary"><h3><a href="#top"><span class="inner">Public Vars</span></a></h3><ul><li class="depth-1"><a href="aerial.bio.utils.files.html#var-canonical-csv-entry-info"><div class="inner"><span>canonical-csv-entry-info</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-check-sto"><div class="inner"><span>check-sto</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-chunk-genome-fnas"><div class="inner"><span>chunk-genome-fnas</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-entry-file-.3Efasta-file"><div class="inner"><span>entry-file-&gt;fasta-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-entry-parts"><div class="inner"><span>entry-parts</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-fastq-.3Efna"><div class="inner"><span>fastq-&gt;fna</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-fastqs-.3Efnas"><div class="inner"><span>fastqs-&gt;fnas</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gaisr-seq-set.3F"><div class="inner"><span>gaisr-seq-set?</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gbank-.3Efna"><div class="inner"><span>gbank-&gt;fna</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gbank-loci-sane-loci"><div class="inner"><span>gbank-loci-sane-loci</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gbanks-.3Efnas"><div class="inner"><span>gbanks-&gt;fnas</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gen-entry-file"><div class="inner"><span>gen-entry-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gen-entry-nv-file"><div class="inner"><span>gen-entry-nv-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gen-name-seq"><div class="inner"><span>gen-name-seq</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gen-name-seq-pairs"><div class="inner"><span>gen-name-seq-pairs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-gen-nc-genome-fnas"><div class="inner"><span>gen-nc-genome-fnas</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-genbank-recs"><div class="inner"><span>genbank-recs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-genbk2gtf"><div class="inner"><span>genbk2gtf</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-get-csv-entry-info"><div class="inner"><span>get-csv-entry-info</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-get-ent-as-csv-info"><div class="inner"><span>get-ent-as-csv-info</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-get-entries"><div class="inner"><span>get-entries</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-get-gaisr-csv-info"><div class="inner"><span>get-gaisr-csv-info</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-get-legacy-csv-info"><div class="inner"><span>get-legacy-csv-info</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-get-selection-fna"><div class="inner"><span>get-selection-fna</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-get-sto-as-csv-info"><div class="inner"><span>get-sto-as-csv-info</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-has-loc.3F"><div class="inner"><span>has-loc?</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-join-sto-fasta-file"><div class="inner"><span>join-sto-fasta-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-join-sto-fasta-lines"><div class="inner"><span>join-sto-fasta-lines</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-make-entry"><div class="inner"><span>make-entry</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-map-aln-seqs"><div class="inner"><span>map-aln-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-map-seqs"><div class="inner"><span>map-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-nms-sqs-.3Efasta-file"><div class="inner"><span>nms-sqs-&gt;fasta-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-print-sto"><div class="inner"><span>print-sto</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-read-aln-seqs"><div class="inner"><span>read-aln-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-read-dirs-aln-seqs"><div class="inner"><span>read-dirs-aln-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-read-seqs"><div class="inner"><span>read-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-reduce-aln-seqs"><div class="inner"><span>reduce-aln-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-reduce-seqs"><div class="inner"><span>reduce-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-sample-fna"><div class="inner"><span>sample-fna</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-sample-fq"><div class="inner"><span>sample-fq</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-seqline-info-mapper"><div class="inner"><span>seqline-info-mapper</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-split-join-fasta-file"><div class="inner"><span>split-join-fasta-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-split-join-ncbi-fasta-file"><div class="inner"><span>split-join-ncbi-fasta-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-sto-.3Ealn"><div class="inner"><span>sto-&gt;aln</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-sto-.3Ealn-blocked"><div class="inner"><span>sto-&gt;aln-blocked</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-sto-.3Efna"><div class="inner"><span>sto-&gt;fna</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-sto-GC-and-seq-lines"><div class="inner"><span>sto-GC-and-seq-lines</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-write-fqrec-to-file"><div class="inner"><span>write-fqrec-to-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-write-fqrecs-to-file"><div class="inner"><span>write-fqrecs-to-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.files.html#var-write-sto"><div class="inner"><span>write-sto</span></div></a></li></ul></div><div class="namespace-docs" id="content"><h1 class="anchor" id="top">aerial.bio.utils.files</h1><div class="doc"><pre class="plaintext">Various bio sequence file format readers, writers, verifiers, and
manipulators.</pre></div><div class="public anchor" id="var-canonical-csv-entry-info"><h3>canonical-csv-entry-info</h3><div class="usage"><code>(canonical-csv-entry-info entries &amp; {:keys [ev], :or {ev 0.0}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-check-sto"><h3>check-sto</h3><div class="usage"><code>(check-sto sto &amp; {printem :printem, :or {printem true}})</code></div><div class="doc"><pre class="plaintext">Checks a sto file to ensure that there are valid characters being
used in the sequences consensus structure line. Will print out
errors in the sto file by sequence number.  Input requires a sto
file</pre></div></div><div class="public anchor" id="var-chunk-genome-fnas"><h3>chunk-genome-fnas</h3><div class="usage"><code>(chunk-genome-fnas genome-fna-dir &amp; {:keys [chunk-size], :or {chunk-size 100}})</code></div><div class="doc"><pre class="plaintext">Take the set of fnas in directory GENOME-FNA-DIR (presumably
created by split-join-ncbi-fasta-file or similar) and aggregate
them into a new smaller set of files, where each new file contains
the contents of CHUNK-SIZE input files (with the possible exception
of the last file having a smaller number).  This is useful for
creating custom data sets for search.
</pre></div></div><div class="public anchor" id="var-entry-file-.3Efasta-file"><h3>entry-file-&gt;fasta-file</h3><div class="usage"><code>(entry-file-&gt;fasta-file efile &amp; {:keys [names-only]})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-entry-parts"><h3>entry-parts</h3><div class="usage"><code>(entry-parts entry &amp; {:keys [ldelta rdelta], :or {ldelta 0, rdelta 0}})</code></div><div class="doc"><pre class="plaintext">ENTRY is a string "name/range/strand", where name is a genome
name, range is of the form start-end and strand is 1 or -1.  At
least name must be supplied.  DELTA is an integer, which will be
subtracted from start and added to end.

Returns a triple [name [start end] strand]
</pre></div></div><div class="public anchor" id="var-fastq-.3Efna"><h3>fastq-&gt;fna</h3><div class="usage"><code>(fastq-&gt;fna fq faot)</code></div><div class="doc"><pre class="plaintext">Convert a fastq format file to a fasta file. Fastq files have 4
lines per 'record' (id, sq, qcid, qc/phred-scores), while the
corresponding fasta has only the id and sq lines per 'record'. Both
FQ and FAOT are file specs.</pre></div></div><div class="public anchor" id="var-fastqs-.3Efnas"><h3>fastqs-&gt;fnas</h3><div class="usage"><code>(fastqs-&gt;fnas fqs &amp; {:keys [outdir]})</code></div><div class="doc"><pre class="plaintext">Convert the fastqs in FQS (a seq) to corresponding fasta files. The
fasta files are named as the fastqs but with file type .fna. If
outdir is given, place the fastas there, otherwise place in same
directory as corresponding fastqs. The items in fqs are file specs!</pre></div></div><div class="public anchor" id="var-gaisr-seq-set.3F"><h3>gaisr-seq-set?</h3><div class="usage"><code>(gaisr-seq-set? x)</code></div><div class="doc"><pre class="plaintext">Returns true if either

X is a filespec string with type extension fna, aln, sto, or gma
X is a java.io.File (presumed to be of one of the above formats)
X is a collection (presumed to have seqences as elements)
</pre></div></div><div class="public anchor" id="var-gbank-.3Efna"><h3>gbank-&gt;fna</h3><div class="usage"><code>(gbank-&gt;fna gbfile fafile)</code></div><div class="doc"><pre class="plaintext">Obtains the genomic sequence (the 'ORIGIN' record) of a genbank
format file gbfile and writes it as the sequence (unblocked) of a
fasta file located in fafile. The id line of the fasta consists of
the 'LOCUS' fields of the gbfile.</pre></div></div><div class="public anchor" id="var-gbank-loci-sane-loci"><h3>gbank-loci-sane-loci</h3><div class="usage"><code>(gbank-loci-sane-loci gbloci)</code></div><div class="doc"><pre class="plaintext">GBFF loci are often, if not incoherent certainly vague and not very
useful. This function takes such and turns them into nice clean loci
with simple start, end and strand. BUG: this will lose some
information in certain cases, but should not with CDS and genes. </pre></div></div><div class="public anchor" id="var-gbanks-.3Efnas"><h3>gbanks-&gt;fnas</h3><div class="usage"><code>(gbanks-&gt;fnas gbfiles outdir)</code></div><div class="doc"><pre class="plaintext">Calls gbank-&gt;fna on all genbank format files in gbfiles, a
collection of fully qualified file specifications of genbank
files. The corresponding fasta file specification is composed of the
basename of a genbank file, with '.fna' file type and placed in
outdir. Outdir is a string which is an output directory
specification, which must exist and be a directory.</pre></div></div><div class="public anchor" id="var-gen-entry-file"><h3>gen-entry-file</h3><div class="usage"><code>(gen-entry-file entries file)</code></div><div class="doc"><pre class="plaintext">Take entries (a collection of nm/s-e/std - see make-entry /
entry-parts) and write them out to file. Returns file</pre></div></div><div class="public anchor" id="var-gen-entry-nv-file"><h3>gen-entry-nv-file</h3><div class="usage"><code>(gen-entry-nv-file entries file)</code></div><div class="doc"><pre class="plaintext">Take entries (a collection of entry [,v]* where v(s) are optional
values (JSD, EVals, etc)) and write them as csv file.</pre></div></div><div class="public anchor" id="var-gen-name-seq"><h3>gen-name-seq</h3><div class="usage"><code>(gen-name-seq entry &amp; {:keys [basedir ldelta rdelta rna], :or {basedir (pams/default-genome-fasta-dir entry), ldelta 0, rdelta 0, rna true}})</code></div><div class="doc"><pre class="plaintext">Generate a pair [entry genome-seq], from ENTRY as possibly modified
by [L|R]DELTA and RNA.  ENTRY is a string "name/range/strand",
where

name is the genome NC name (and we only currently support NCs),

range is of the form start-end, where start and end are integers (1
based...) for the start and end coordinates of name's sequence to
return.  NOTE: start &lt; end, as reverse compliment information comes
from strand.

strand is either -1 for reverse compliment or 1 for standard 5'-&gt;3'

LDELTA is an integer, which will be _subtracted_ from start.  So,
ldelta &lt; 0 _removes_ |ldelta| bases from 5', ldelta &gt; 0 'tacks on'
ldelta extra bases to the 5' end.  Defaults to 0 (no change).

RDELTA is an integer, which will be _added_ to the end.  So, rdelta
&lt; 0 _removes_ |rdelta| bases from 3', rdelta &gt; 0 'tacks on' rdelta
extra bases to the 3' end.  Defaults to 0 (no change).

If RNA is true, change Ts to Us, otherwise return unmodified
sequence.

BASEDIR is the location of the NC fasta files.  Generally, this
should always be the default location.
</pre></div></div><div class="public anchor" id="var-gen-name-seq-pairs"><h3>gen-name-seq-pairs</h3><div class="usage"><code>(gen-name-seq-pairs entries &amp; {:keys [basedir strand ldelta rdelta rna], :or {basedir pams/default-genome-fasta-dir, strand 0, ldelta 0, rdelta 0, rna true}})</code></div><div class="doc"><pre class="plaintext">Generate a sequence of pairs [name genome-seq] from the given
collection denoted by entries, which is either a collection or a
string denoting an entry filespec (either hand made or via
gen-entry-file or similar).

See gen-name-seq for details.  Basically this is (map gen-name-seq
entries) with some options.  In particular, entries may lack a
strand component (again see gen-name-seq for format details), and
STRAND here would be used to indicate all entries are to have this
strand.  If entries have a strand component, strand should be 0,
otherwise you will force all entries to the strand given.  Strand
is either -1 (reverse compliment) or 1 for standard 5'-&gt;3', or 0
meaning 'use strand component in entries'

BASEDIR is the location of the NC fasta files.  Generally, this
should always be the default location.
</pre></div></div><div class="public anchor" id="var-gen-nc-genome-fnas"><h3>gen-nc-genome-fnas</h3><div class="usage"><code>(gen-nc-genome-fnas full-nc-genomes-filespec)</code></div><div class="doc"><pre class="plaintext">One shot genome sequence fnas generator.  Typically used once per
data update.  Needs to be generalized to be able to use Genbank fna
archives.  Currently assumes one fna AND ONLY NC_* genomes.
</pre></div></div><div class="public anchor" id="var-genbank-recs"><h3>genbank-recs</h3><div class="usage"><code>(genbank-recs gbfile &amp; {:keys [feats attrs], :or {feats ["gene" "CDS" "misc_feature" "rRNA" "tRNA" "tmRNA" "ncRNA"], attrs ["gene" "locus_tag" "old_locus_tag" "protein_id" "condon_start" "gene_synonym" "db_xref"]}})</code></div><div class="doc"><pre class="plaintext">Takes a genbank file, gbfile - a string file specification, and a
set of features (feats) and associated attributes (attrs),so called
qualifiers and values, and returns a vector where the first element
describes the LOCUS of the gbfile and each subsequent element is a
triple [feature loci attr-map], where feature (a string) is a one of
feats, looci is a vector [start end strand], (all strings, strand =
1|-1), and attr-map is a map with keys in attrs and values, the
associated values from the genbank record.</pre></div></div><div class="public anchor" id="var-genbk2gtf"><h3>genbk2gtf</h3><div class="usage"><code>(genbk2gtf gbfile gtfout)</code></div><div class="doc"><pre class="plaintext">Extract a GTF (gene transfer format) file from the record
information (features and their attributes) given in the genbank
format file gbfile. The GTF file also includes the p_id attribute on
CDS records, as required by the cuff* suite of software. For gene_id
and transcript_id, uses the first from the ordered list:
old_locus_tag, locus_tag, gene, protein_id. </pre></div></div><div class="public anchor" id="var-get-csv-entry-info"><h3>get-csv-entry-info</h3><div class="usage"><code>(get-csv-entry-info csv-hit-file)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-ent-as-csv-info"><h3>get-ent-as-csv-info</h3><div class="usage"><code>(get-ent-as-csv-info ent-file)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-entries"><h3>get-entries</h3><div class="usage"><code>(get-entries filespec &amp; [seqs])</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-gaisr-csv-info"><h3>get-gaisr-csv-info</h3><div class="usage"><code>(get-gaisr-csv-info rows)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-legacy-csv-info"><h3>get-legacy-csv-info</h3><div class="usage"><code>(get-legacy-csv-info rows)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-selection-fna"><h3>get-selection-fna</h3><div class="usage"><code>(get-selection-fna selections)</code></div><div class="doc"><pre class="plaintext">Get a fasta file for SELECTIONS.  If selections is a file, assumes
it is in fact the fasta file.  If selections is a collection,
assumes the collection is a set of pairs [nm sq], and converts to
corresponding fasta file.  Returns full path of result file.</pre></div></div><div class="public anchor" id="var-get-sto-as-csv-info"><h3>get-sto-as-csv-info</h3><div class="usage"><code>(get-sto-as-csv-info stofile &amp; {:keys [ev], :or {ev 0.0}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-has-loc.3F"><h3>has-loc?</h3><div class="usage"><code>(has-loc? entries)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-join-sto-fasta-file"><h3>join-sto-fasta-file</h3><div class="usage"><code>(join-sto-fasta-file in-filespec out-filespec &amp; {origin :origin, :or {origin ""}})</code></div><div class="doc"><pre class="plaintext">Joins (de-blocks) unblocked sequence lines in a sto file or fasta
file. If in-filespec is a sto file, ORIGIN is a #=GF line indicating
tool origin of file.  For example, '#=GF AU Infernal 1.0.2'. For
stos defaults to nothing, for fastas, not used.</pre></div></div><div class="public anchor" id="var-join-sto-fasta-lines"><h3>join-sto-fasta-lines</h3><div class="usage"><code>(join-sto-fasta-lines infilespec origin)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-make-entry"><h3>make-entry</h3><div class="usage"><code>(make-entry evec)</code><code>(make-entry nm s e st)</code></div><div class="doc"><pre class="plaintext">'Inverse' of entry-parts.  EVEC is a vector of shape [nm [s e]
strd], where nm is the entry name, S and E are the start and end
coordinates in the genome, and strd is the strand marker, 1 or
-1. Returns the full entry as: nm/s-e/strd
</pre></div></div><div class="public anchor" id="var-map-aln-seqs"><h3>map-aln-seqs</h3><div class="usage"><code>(map-aln-seqs f cols filespec)</code><code>(map-aln-seqs f par cols filespec &amp; filespecs)</code></div><div class="doc"><pre class="plaintext">
</pre></div></div><div class="public anchor" id="var-map-seqs"><h3>map-seqs</h3><div class="usage"><code>(map-seqs f filespec)</code><code>(map-seqs f par filespec &amp; filespecs)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-nms-sqs-.3Efasta-file"><h3>nms-sqs-&gt;fasta-file</h3><div class="usage"><code>(nms-sqs-&gt;fasta-file nms-sqs filespec)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-print-sto"><h3>print-sto</h3><div class="usage"><code>(print-sto seq-lines structure)</code></div><div class="doc"><pre class="plaintext">takes sequence lines and a structure line and writes it into a sto
format file. the seq-lines needs to be a collection of [name
sequence] pairs. structure is a string. Simply prints out to the
repl.</pre></div></div><div class="public anchor" id="var-read-aln-seqs"><h3>read-aln-seqs</h3><div class="usage"><code>(read-aln-seqs filespec &amp; {cols :cols, :or {cols false}})</code></div><div class="doc"><pre class="plaintext">Read the _aligned_ sequences in FILESPEC and return them in a
Clojure seq.  Filespec can denote either an aln, sto, or gma file
format file.  If COLS is true, return the _columns_ of the
alignment (including gap characters).
</pre></div></div><div class="public anchor" id="var-read-dirs-aln-seqs"><h3>read-dirs-aln-seqs</h3><div class="usage"><code>(read-dirs-aln-seqs dir &amp; {dirdir :dirdir, cols :cols, ftypes :ftypes, :or {dirdir false, cols false, ftypes ["sto"]}})</code></div><div class="doc"><pre class="plaintext">Apply read-aln-seqs across FTYPES in DIR.  FTYPES is a vector of
one or more file formats (as type extensions) taken from #{"fna",
"aln", "sto", "gma"}, producing a seq of sequence sets, each
set being a seq of the sequences in a matching file.  NOTE: each
such set can be viewed as a 'matrix' of the elements (bases/gaps)
of the sequences.

If COLS is true, return the transpose of each obtained sequence
'matrix', i.e., return the seq of column seqs.  NOTE: as the
formats are alignment formats, all sequences in a file are of the
same length (including gaps).

If DIRDIR is true, dir is taken as a directory of directories, each
of which will have read-aln-seqs applied to it per the above
description and the return will be a seq of all such applications.

So, if dirdir is false, the result will have the form:

(seqs-from-file1 ... seqs-from-filen), where filei is in dir

If dirdir is true, the result will be nested one level more:

((seqs-from-dir1-file1 ... seqs-from-dir1-filek)
 ...
 (seqs-from-dirn-file1 ... seqs-from-dirn-filel))
</pre></div></div><div class="public anchor" id="var-read-seqs"><h3>read-seqs</h3><div class="usage"><code>(read-seqs input &amp; {:keys [info ftype], :or {info :data}})</code></div><div class="doc"><pre class="plaintext">Read the sequences in FILESPEC and return set as a lazy seq.
Filespec can denote either a fna, fa, hitfna, aln, sto, or gma file
format file.
</pre></div></div><div class="public anchor" id="var-reduce-aln-seqs"><h3>reduce-aln-seqs</h3><div class="usage"><code>(reduce-aln-seqs f fr cols filespecs)</code><code>(reduce-aln-seqs f fr v cols filespecs)</code></div><div class="doc"><pre class="plaintext">
</pre></div></div><div class="public anchor" id="var-reduce-seqs"><h3>reduce-seqs</h3><div class="usage"><code>(reduce-seqs f fr filespecs)</code><code>(reduce-seqs f fr v filespecs)</code></div><div class="doc"><pre class="plaintext">
</pre></div></div><div class="public anchor" id="var-sample-fna"><h3>sample-fna</h3><div class="usage"><code>(sample-fna p f)</code><code>(sample-fna p f sampfna)</code></div><div class="doc"><pre class="plaintext">Sample the sequences in f, a fasta file, with probability
p. Returns a seq of pairs suitable for writing a fasta file: [id,
sq]. The id is the corresponding id of the sq in f. In the 3 arg
case, sampfna is a filespec for an output fasta file where the
sampling is written.</pre></div></div><div class="public anchor" id="var-sample-fq"><h3>sample-fq</h3><div class="usage"><code>(sample-fq p f)</code><code>(sample-fq p f sampfq)</code></div><div class="doc"><pre class="plaintext">Sample the sequences in f, a fastq file, with probability
p. Returns a seq of quadtuples suitable for writing a fastq
file: [id, sq, qcdesc qc]. The id is the corresponding id of the sq
in f. qcdesc and qc are the corresponding quality description line
and the quality score line. In the 3 arg case, sampfq is a filespec
for an output fastq file where the sampling is written.</pre></div></div><div class="public anchor" id="var-seqline-info-mapper"><h3>seqline-info-mapper</h3><div class="usage"><code>(seqline-info-mapper type info)</code></div><div class="doc"><pre class="plaintext">Helper function for READ-SEQS.  Returns the function to map over
seq lines to obtain the requested info.  TYPE is supported seq file
type (aln, sto, fna, fa, gma).  INFO is either :name for the
sequence identifier, :data for the sequence data, or :both for name
and data.

Impl Note: while this almost begs for multimethods, that would
actually increase the complexity as it would mean 14 methods to
cover the cases...
</pre></div></div><div class="public anchor" id="var-split-join-fasta-file"><h3>split-join-fasta-file</h3><div class="usage"><code>(split-join-fasta-file in-file &amp; {:keys [base pat namefn entryfn testfn], :or {base "", pat #"^&gt;gi", entryfn identity, testfn (fn [x y] true)}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-split-join-ncbi-fasta-file"><h3>split-join-ncbi-fasta-file</h3><div class="usage"><code>(split-join-ncbi-fasta-file in-file)</code></div><div class="doc"><pre class="plaintext">Split a fasta file IN-FILE into the individual sequences and
unblock the sequence if blocked.  The resulting individual [nm sq]
pairs are written to files named for the NC name in the gi line of
in-file and in the DEFAULT-GENOME-FASTA-DIR location.

The main use of this function is to take a refseq fasta
db (composed of many multi seq fasta files) and split the db into a
normed set of named sequence files for quick access to sequence per
name in various other processing (see gen-name-seq for example).

Canonical use case example:

(fs/dodir "/data2/BioData/Fasta" ; RefSeqxx fasta files
          #(fs/directory-files % "fna")
          #(split-join-ncbi-fasta-file %))
</pre></div></div><div class="public anchor" id="var-sto-.3Ealn"><h3>sto-&gt;aln</h3><div class="usage"><code>(sto-&gt;aln stoin alnout &amp; {blocked :blocked, :or {blocked false}})</code></div><div class="doc"><pre class="plaintext">Convert a stockhom format alignment file into its ClustalW
equivalent ALN format.  STOIN is the filespec for the stockholm
format file and ALNOUT is the filespec for the resulting
conversion (it is overwritten if it already exists!)

BLOCKED is a boolean indicating whether the output should be
blocked (60 chars per chunk).  Default is unblocked.</pre></div></div><div class="public anchor" id="var-sto-.3Ealn-blocked"><h3>sto-&gt;aln-blocked</h3><div class="usage"><code>(sto-&gt;aln-blocked stoin alnout)</code></div><div class="doc"><pre class="plaintext">Convert a stockhom format alignment file into its ClustalW
equivalent BLOCKED ALN format. Blocking is done in 60 character
chunks.  STOIN is the filespec for the stockholm format file and
ALNOUT is the filespec for the resulting conversion (it is
overwritten if it already exists!)</pre></div></div><div class="public anchor" id="var-sto-.3Efna"><h3>sto-&gt;fna</h3><div class="usage"><code>(sto-&gt;fna stoin fnaout)</code></div><div class="doc"><pre class="plaintext">Convert a sto file into a fasta file.  Split seq lines into names
and seq data and interleave these.  Seq data has all gap characters
removed.</pre></div></div><div class="public anchor" id="var-sto-GC-and-seq-lines"><h3>sto-GC-and-seq-lines</h3><div class="usage"><code>(sto-GC-and-seq-lines stofilespec)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-write-fqrec-to-file"><h3>write-fqrec-to-file</h3><div class="usage"><code>(write-fqrec-to-file ot rec)</code></div><div class="doc"><pre class="plaintext">Write a fastq 'record' to a file.  OT is an output file
descriptor (an already opened output-stream writer).  REC is a
vector quad [id sq aux qc], representing the id line, the sequence
line, the auxilliary information line and the quality control line
for a fastq format file.</pre></div></div><div class="public anchor" id="var-write-fqrecs-to-file"><h3>write-fqrecs-to-file</h3><div class="usage"><code>(write-fqrecs-to-file ot recs)</code></div><div class="doc"><pre class="plaintext">Write fastq 'records' to file. OT is an output file descriptor (an
already opened output-stream writer). RECS is a vector/sequence of
quads [id sq aux qc], each representing the id line, the sequence
line, the auxilliary information line and the quality control line
for a fastq format file.</pre></div></div><div class="public anchor" id="var-write-sto"><h3>write-sto</h3><div class="usage"><code>(write-sto newsto auth-lines comment-lines nm-sq-pairs ss-lines)</code></div><div class="doc"><pre class="plaintext">A work in progress...  Write a new sto composed of the various
given parts to the file spec given as NEWSTO.  AUTH-LINES are the
authoring header lines - including the STOCKHOLM line.  Generally
there are two of these - the STOCKHOLM line (with version) and the
originating author or program that generated the content (for
example, Infernal).

COMMENT-LINES is a collection of the #=GF/GC lines, with the
exception of the GC SS_cons and RF lines.  Comment-lines may be
empty (for example, []).

NM-SQ-PAIRS is a collection (typically vector/list) of pairs of the
entries (name/start-end/strand) and the associated sequence (in
gapped form).  If this is created via JOIN-STO-FASTA-LINES, the
vector of [id sq] pairs that is the sequence part of the
nm-sq-pair, will have the id part filtered out automatically.

SS-LINES is the set of 'secondary structure' lines.  These are the
GC SS_cons and RF lines.  SS-LINES may contain the final '//' line
or not.  If not, it is still written to the file, if so, only the
one '//' is written.
</pre></div></div></div></body></html>