<!DOCTYPE html>
<html><head><meta charset="UTF-8"><link href="css/default.css" rel="stylesheet" type="text/css"><script src="js/jquery.min.js" type="text/javascript"></script><script src="js/page_effects.js" type="text/javascript"></script><title>aerial.bio.utils.seqxform documentation</title></head><body><div id="header"><h2>Generated by <a href="https://github.com/weavejester/codox">Codox</a></h2><h1><a href="index.html"><span class="project-title"><span class="project-name">Aerial.bio.utils</span> <span class="project-version">1.1.0</span></span></a></h1></div><div class="sidebar primary"><ul class="index-link"><li class="depth-1 "><a href="index.html"><div class="inner">Index</div></a></li></ul><h3 class="no-link"><span class="inner">Namespaces</span></h3><ul><li class="depth-1"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>aerial</span></div></div></li><li class="depth-2"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>bio</span></div></div></li><li class="depth-3"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>utils</span></div></div></li><li class="depth-4 branch"><a href="aerial.bio.utils.aligners.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>aligners</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.extools.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>extools</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.files.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>files</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.filters.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>filters</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.infoth.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>infoth</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.params.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>params</span></div></a></li><li class="depth-4 current"><a href="aerial.bio.utils.seqxform.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>seqxform</span></div></a></li></ul></div><div class="sidebar secondary"><h3><a href="#top"><span class="inner">Public Vars</span></a></h3><ul><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BAMINO-ACID-MAP.2B"><div class="inner"><span>+AMINO-ACID-MAP+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BDNA-CONDON-MAP.2B"><div class="inner"><span>+DNA-CONDON-MAP+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BIUPAC.2B"><div class="inner"><span>+IUPAC+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BNONSTD-RNA.2B"><div class="inner"><span>+NONSTD-RNA+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BREVERSE-AMINO-ACID-MAP.2B"><div class="inner"><span>+REVERSE-AMINO-ACID-MAP+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BREVERSE-DNA-CODON-MAP.2B"><div class="inner"><span>+REVERSE-DNA-CODON-MAP+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BREVERSE-RNA-CODON-MAP.2B"><div class="inner"><span>+REVERSE-RNA-CODON-MAP+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BRNA-CONDON-MAP.2B"><div class="inner"><span>+RNA-CONDON-MAP+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BRY-XLATE.2B"><div class="inner"><span>+RY-XLATE+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BSTART-CODON.2B"><div class="inner"><span>+START-CODON+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-.2BSTOP-CODONS.2B"><div class="inner"><span>+STOP-CODONS+</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-alphabet"><div class="inner"><span>alphabet</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-check-limit"><div class="inner"><span>check-limit</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-cut-cols"><div class="inner"><span>cut-cols</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-degap-seqs"><div class="inner"><span>degap-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-dint-seq-shuffle"><div class="inner"><span>dint-seq-shuffle</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-dint-shuffle"><div class="inner"><span>dint-shuffle</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-filter-pgap"><div class="inner"><span>filter-pgap</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-gap-count"><div class="inner"><span>gap-count</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-gap-percent"><div class="inner"><span>gap-percent</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-gaps.3F"><div class="inner"><span>gaps?</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-gen-random-bioseq"><div class="inner"><span>gen-random-bioseq</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-legal-ends"><div class="inner"><span>legal-ends</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-limit-counter"><div class="inner"><span>limit-counter</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-monte-carlo-subseq-cnts"><div class="inner"><span>monte-carlo-subseq-cnts</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-monte-carlo-subseq-cnts-body"><div class="inner"><span>monte-carlo-subseq-cnts-body</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-monte-carlo-subseq-freq"><div class="inner"><span>monte-carlo-subseq-freq</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-norm-elements"><div class="inner"><span>norm-elements</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-nt-cntn"><div class="inner"><span>nt-cntn</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-nts-map"><div class="inner"><span>nts-map</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-ntseq-.3Eaaseq"><div class="inner"><span>ntseq-&gt;aaseq</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-pmonte-carlo-subseq-cnts"><div class="inner"><span>pmonte-carlo-subseq-cnts</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-print-limit-info"><div class="inner"><span>print-limit-info</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-reduce-seqcnt-results"><div class="inner"><span>reduce-seqcnt-results</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-reverse-compliment"><div class="inner"><span>reverse-compliment</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-seqXlate"><div class="inner"><span>seqXlate</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-shuffles"><div class="inner"><span>shuffles</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.seqxform.html#var-test-nts"><div class="inner"><span>test-nts</span></div></a></li></ul></div><div class="namespace-docs" id="content"><h1 class="anchor" id="top">aerial.bio.utils.seqxform</h1><div class="doc"><pre class="plaintext">Various sequence transformation and generation functions.
</pre></div><div class="public anchor" id="var-.2BAMINO-ACID-MAP.2B"><h3>+AMINO-ACID-MAP+</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-.2BDNA-CONDON-MAP.2B"><h3>+DNA-CONDON-MAP+</h3><div class="usage"></div><div class="doc"><pre class="plaintext">
</pre></div></div><div class="public anchor" id="var-.2BIUPAC.2B"><h3>+IUPAC+</h3><div class="usage"></div><div class="doc"><pre class="plaintext">These are the codes for &quot;bases&quot; used in alignments
</pre></div></div><div class="public anchor" id="var-.2BNONSTD-RNA.2B"><h3>+NONSTD-RNA+</h3><div class="usage"></div><div class="doc"><pre class="plaintext">These are the codes for all non standard &quot;bases&quot; used in
alignments</pre></div></div><div class="public anchor" id="var-.2BREVERSE-AMINO-ACID-MAP.2B"><h3>+REVERSE-AMINO-ACID-MAP+</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-.2BREVERSE-DNA-CODON-MAP.2B"><h3>+REVERSE-DNA-CODON-MAP+</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-.2BREVERSE-RNA-CODON-MAP.2B"><h3>+REVERSE-RNA-CODON-MAP+</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-.2BRNA-CONDON-MAP.2B"><h3>+RNA-CONDON-MAP+</h3><div class="usage"></div><div class="doc"><pre class="plaintext">
</pre></div></div><div class="public anchor" id="var-.2BRY-XLATE.2B"><h3>+RY-XLATE+</h3><div class="usage"></div><div class="doc"><pre class="plaintext">Map for translating DNA and RNA seqs to purine and pyrimidine seqs.
Assumes normed-elements (for which see)</pre></div></div><div class="public anchor" id="var-.2BSTART-CODON.2B"><h3>+START-CODON+</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-.2BSTOP-CODONS.2B"><h3>+STOP-CODONS+</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-alphabet"><h3>alphabet</h3><div class="usage"><code>(alphabet type)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-check-limit"><h3>check-limit</h3><div class="usage"><code>(check-limit start len ss newseq nts-map)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-cut-cols"><h3>cut-cols</h3><div class="usage"><code>(cut-cols sq secol)</code><code>(cut-cols sq s e)</code></div><div class="doc"><pre class="plaintext">Cut bases in columns from sequence SQ
</pre></div></div><div class="public anchor" id="var-degap-seqs"><h3>degap-seqs</h3><div class="usage"><code>(degap-seqs seqs)</code><code>(degap-seqs seqs gap-chars)</code></div><div class="doc"><pre class="plaintext">Remove gap characters from a sequence or set of sequences.  These
would be from an alignment set.  It is not clear how / where useful
this is/would be as it destroys the alignment spacing.  Other than
a single sequence situation, use degap-tuples instead!!

Gap chars are either the defaults \. \- or those in gap-chars (a
seqable collection).
</pre></div></div><div class="public anchor" id="var-dint-seq-shuffle"><h3>dint-seq-shuffle</h3><div class="usage"><code>(dint-seq-shuffle limit s)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-dint-shuffle"><h3>dint-shuffle</h3><div class="usage"><code>(dint-shuffle start len nts-map legal-ends ss newseq &amp; base-starters)</code></div><div class="doc"><pre class="plaintext">THIS NEEDS TO BE REDONE!  NOTE IT CAN&apos;T BE CALLED OUTSIDE OF WHERE
IT IS CALLED IN dint-seq-shuffle, DUE TO EGREGIOUSLY POOR GLOBAL
BINDINGS AND SUCH.
</pre></div></div><div class="public anchor" id="var-filter-pgap"><h3>filter-pgap</h3><div class="usage"><code>(filter-pgap seqs pgap)</code><code>(filter-pgap seqs pgap gap-chars)</code></div><div class="doc"><pre class="plaintext">Filter the sequence set seqs, by returning only those that have
less than a pgap percentage of gap characters. Gap chars are either
the defaults \. \- or those in gap-chars (a seqable collection).</pre></div></div><div class="public anchor" id="var-gap-count"><h3>gap-count</h3><div class="usage"><code>(gap-count sq)</code><code>(gap-count sq gap-chars)</code></div><div class="doc"><pre class="plaintext">Return the number of &quot;gap characters&quot; in sq.  For the single
parameter case gap chars are taken as, a \. or \-.  For the
gap-chars case gap characters are the elements (characters) in
gap-chars (a seqable collection).
</pre></div></div><div class="public anchor" id="var-gap-percent"><h3>gap-percent</h3><div class="usage"><code>(gap-percent sq)</code><code>(gap-percent sq gap-chars)</code></div><div class="doc"><pre class="plaintext">Return the percentage of sq (typically an aligned variant of a
genome sequence) that is comprised of gap characters.  For the
single parameter case, gaps are taken as \. and \-.  For the
gap-chars case, gap characters are the elements (characters) in
gap-chars (a seqable collection).</pre></div></div><div class="public anchor" id="var-gaps.3F"><h3>gaps?</h3><div class="usage"><code>(gaps? k)</code><code>(gaps? k gap-chars)</code></div><div class="doc"><pre class="plaintext">Return whether K, a char, string, or coll, contains a &quot;gap
character&quot;.  For the single parameter case gap chars are taken as,
a \. or \-.  For the gap-chars case gap characters are the
elements (characters) in gap-chars (a seqable collection).
</pre></div></div><div class="public anchor" id="var-gen-random-bioseq"><h3>gen-random-bioseq</h3><div class="usage"><code>(gen-random-bioseq type size &amp; seed)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-legal-ends"><h3>legal-ends</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-limit-counter"><h3>limit-counter</h3><h4 class="dynamic">dynamic</h4><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-monte-carlo-subseq-cnts"><h3>monte-carlo-subseq-cnts</h3><div class="usage"><code>(monte-carlo-subseq-cnts &amp; {:keys [type size sseq sample-size seed], :or {:type :dna, :size 1024, :sseq &quot;&quot;, :sample-size 10000, :seed nil}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-monte-carlo-subseq-cnts-body"><h3>monte-carlo-subseq-cnts-body</h3><div class="usage"><code>(monte-carlo-subseq-cnts-body type sample-size seqsize sseq len seed)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-monte-carlo-subseq-freq"><h3>monte-carlo-subseq-freq</h3><div class="usage"><code>(monte-carlo-subseq-freq &amp; {:keys [type size sseq sample-size seed par], :or {:type :dna, :size 1024, :sseq &quot;&quot;, :sample-size 10000, :seed nil, :par 10}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-norm-elements"><h3>norm-elements</h3><div class="usage"><code>(norm-elements seqs)</code></div><div class="doc"><pre class="plaintext">&quot;Normalize&quot; elements in sequences by ensuring each character is
mapped to its uppercase variant.</pre></div></div><div class="public anchor" id="var-nt-cntn"><h3>nt-cntn</h3><div class="usage"><code>(nt-cntn n sequence)</code></div><div class="doc"><pre class="plaintext">Frequencies of N-(nucleotides/amino-acids) of sequence SEQUENCE,
i.e., for N=2 and bases freq of dinucleotides.  Effectively a
rename of utils:freqn.
</pre></div></div><div class="public anchor" id="var-nts-map"><h3>nts-map</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-ntseq-.3Eaaseq"><h3>ntseq-&gt;aaseq</h3><div class="usage"><code>(ntseq-&gt;aaseq ntseq &amp; {:keys [stop rna-dna], :or {stop true, rna-dna :rna}})</code></div><div class="doc"><pre class="plaintext">Translate the nucleotide sequence NTSEQ into its corresponding
amino acid (protein) sequence.  STOP true implies ntseq has its
stop codon at the end, otherwise ntseq is assumed to not have its
stop codon.  Defaults to true.

RNA-DNA = :rna implies that ntseq is given as rna (with uracil
bases); = :dna implies ntseq given as dna (with thymine bases).
Defaults to :rna

Precondition: ntseq must have a multiple of 3 bases!!

Returns the amino acid (protein) sequence for ntseq (sans start
stop codon translations, if they were included)
</pre></div></div><div class="public anchor" id="var-pmonte-carlo-subseq-cnts"><h3>pmonte-carlo-subseq-cnts</h3><div class="usage"><code>(pmonte-carlo-subseq-cnts &amp; {:keys [type size sseq sample-size seed par], :or {:type :dna, :size 1024, :sseq &quot;&quot;, :sample-size 10000, :seed nil, :par 10}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-print-limit-info"><h3>print-limit-info</h3><h4 class="dynamic">dynamic</h4><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-reduce-seqcnt-results"><h3>reduce-seqcnt-results</h3><div class="usage"><code>(reduce-seqcnt-results cntmaps)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-reverse-compliment"><h3>reverse-compliment</h3><div class="usage"><code>(reverse-compliment sq)</code></div><div class="doc"><pre class="plaintext">Reverse compliment of DNA/RNA sequence SQ. Return the sequence that
would pair with (reverse SQ).

Ex: (reverse-compliment &quot;AAGGAAUUCC&quot;) =&gt; &quot;GGAAUUCCUU&quot;
</pre></div></div><div class="public anchor" id="var-seqXlate"><h3>seqXlate</h3><div class="usage"><code>(seqXlate sqs &amp; {:keys [xmap], :or {xmap +RY-XLATE+}})</code></div><div class="doc"><pre class="plaintext">Translate sqs according to the translation map xmap.  SQS is either
a single sequence (string) or a collection of such.  It cannot be a
file.  Each sequence has its elements transformed according xmap
and each is returned as a string.
</pre></div></div><div class="public anchor" id="var-shuffles"><h3>shuffles</h3><h4 class="dynamic">dynamic</h4><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-test-nts"><h3>test-nts</h3><div class="usage"></div><div class="doc"><pre class="plaintext"></pre></div></div></div></body></html>