<!DOCTYPE html>
<html><head><meta charset="UTF-8"><link href="css/default.css" rel="stylesheet" type="text/css"><script src="js/jquery.min.js" type="text/javascript"></script><script src="js/page_effects.js" type="text/javascript"></script><title>aerial.bio.utils.extools documentation</title></head><body><div id="header"><h2>Generated by <a href="https://github.com/weavejester/codox">Codox</a></h2><h1><a href="index.html"><span class="project-title"><span class="project-name">Aerial.bio.utils</span> <span class="project-version">2.0.0</span></span></a></h1></div><div class="sidebar primary"><ul class="index-link"><li class="depth-1 "><a href="index.html"><div class="inner">Index</div></a></li></ul><h3 class="no-link"><span class="inner">Topics</span></h3><ul><li class="depth-1 "><a href="intro.html"><div class="inner"><span>Introduction to bio.utils</span></div></a></li></ul><h3 class="no-link"><span class="inner">Namespaces</span></h3><ul><li class="depth-1"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>aerial</span></div></div></li><li class="depth-2"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>bio</span></div></div></li><li class="depth-3"><div class="no-link"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>utils</span></div></div></li><li class="depth-4 branch"><a href="aerial.bio.utils.aligners.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>aligners</span></div></a></li><li class="depth-4 branch current"><a href="aerial.bio.utils.extools.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>extools</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.files.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>files</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.filters.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>filters</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.infoth.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>infoth</span></div></a></li><li class="depth-4 branch"><a href="aerial.bio.utils.params.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>params</span></div></a></li><li class="depth-4"><a href="aerial.bio.utils.seqxform.html"><div class="inner"><span class="tree"><span class="top"></span><span class="bottom"></span></span><span>seqxform</span></div></a></li></ul></div><div class="sidebar secondary"><h3><a href="#top"><span class="inner">Public Vars</span></a></h3><ul><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-assert-tools-exist"><div class="inner"><span>assert-tools-exist</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-blast"><div class="inner"><span>blast</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-blastn"><div class="inner"><span>blastn</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-blastpgm"><div class="inner"><span>blastpgm</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-canda"><div class="inner"><span>canda</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-candf"><div class="inner"><span>candf</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cands"><div class="inner"><span>cands</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cd-hit-est"><div class="inner"><span>cd-hit-est</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cmalign"><div class="inner"><span>cmalign</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cmbuild"><div class="inner"><span>cmbuild</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cmcalibrate"><div class="inner"><span>cmcalibrate</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cmf-post-process-combine"><div class="inner"><span>cmf-post-process-combine</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cmf-post-process-filter"><div class="inner"><span>cmf-post-process-filter</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cmfinder"><div class="inner"><span>cmfinder</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cmfinder*"><div class="inner"><span>cmfinder*</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-cmsearch"><div class="inner"><span>cmsearch</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-correct-dir-stos"><div class="inner"><span>correct-dir-stos</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-correct-sto-coordinates"><div class="inner"><span>correct-sto-coordinates</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-embl-sto-.3Enc-sto"><div class="inner"><span>embl-sto-&gt;nc-sto</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-embl-to-nc"><div class="inner"><span>embl-to-nc</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-aln-sqs"><div class="inner"><span>entry-aln-sqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-file-.3Eblastdb"><div class="inner"><span>entry-file-&gt;blastdb</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-file-difference"><div class="inner"><span>entry-file-difference</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-file-difference-with-seqs"><div class="inner"><span>entry-file-difference-with-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-file-intersect"><div class="inner"><span>entry-file-intersect</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-file-intersect-with-seqs"><div class="inner"><span>entry-file-intersect-with-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-file-setop"><div class="inner"><span>entry-file-setop</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-file-union"><div class="inner"><span>entry-file-union</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-file-union-with-seqs"><div class="inner"><span>entry-file-union-with-seqs</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-entry-range-.3Efasta-file"><div class="inner"><span>entry-range-&gt;fasta-file</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-filter-base-hits-to-most-likely"><div class="inner"><span>filter-base-hits-to-most-likely</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-fuzzy-entry-file-difference"><div class="inner"><span>fuzzy-entry-file-difference</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-fuzzy-entry-file-intersect"><div class="inner"><span>fuzzy-entry-file-intersect</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-fuzzy-entry-file-union"><div class="inner"><span>fuzzy-entry-file-union</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-genome-seq-dup.3F"><div class="inner"><span>genome-seq-dup?</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-get-blast-id-info-new"><div class="inner"><span>get-blast-id-info-new</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-get-cd-hit-infile"><div class="inner"><span>get-cd-hit-infile</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-get-cmfsto-glob"><div class="inner"><span>get-cmfsto-glob</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-get-embl-blast-candidates"><div class="inner"><span>get-embl-blast-candidates</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-get-entry-set"><div class="inner"><span>get-entry-set</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-get-tool-path"><div class="inner"><span>get-tool-path</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-infernal-2.2B.3F"><div class="inner"><span>infernal-2+?</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-mk-infernal-cmd"><div class="inner"><span>mk-infernal-cmd</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-parse-blast"><div class="inner"><span>parse-blast</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-process-blast-for-cmfinder"><div class="inner"><span>process-blast-for-cmfinder</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-rank-em"><div class="inner"><span>rank-em</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-refseq-entry.3F"><div class="inner"><span>refseq-entry?</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-summary-map"><div class="inner"><span>summary-map</span></div></a></li><li class="depth-1"><a href="aerial.bio.utils.extools.html#var-tblastn"><div class="inner"><span>tblastn</span></div></a></li></ul></div><div class="namespace-docs" id="content"><h1 class="anchor" id="top">aerial.bio.utils.extools</h1><div class="doc"><pre class="plaintext">Intended to include/contain various utilities for operating with
and on sequence data.  Originally this was all about working
with (NCBI) Blast+, CMFINDER (and associated utils) and Infernal
tools, but now may branch out to many others.</pre></div><div class="public anchor" id="var-assert-tools-exist"><h3>assert-tools-exist</h3><div class="usage"><code>(assert-tools-exist tool-vec)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-blast"><h3>blast</h3><div class="usage"><code>(blast pgm in &amp; k__1538__auto__)</code></div><div class="doc"><pre class="plaintext">Implementation of blast requests.  See blastn, tblastn, etc. for details
on main arguments.  PGM is a keyword denoting which blast program.</pre></div></div><div class="public anchor" id="var-blastn"><h3>blastn</h3><div class="usage"><code>(blastn in &amp; kw-args)</code></div><div class="doc"><pre class="plaintext">NCBI blastn (from the blast+ toolset).

IN is the input filespec in fasta format.

KEYWORD ARGS:
:out is the output filespec for direct bast output, default `in`.blast
:blastdb the filespec for the preformatted blast database to use
:strand (string) plus or minus, defaults to plus
:word-size blast bp slice size (chunk size for alignment), default 8
:evalue (real) expectation (E) value threshold for saving hits, default 10
:fmt NCBI blastn out fmt string, defaults to:
     10 qseqid qstart qend evalue sseqid sstart send
:fn post processing fn of at least 1 arg for blast output, eg parse-blast
:fnargs any other args for fn, e.g., output filespec</pre></div></div><div class="public anchor" id="var-blastpgm"><h3>blastpgm</h3><div class="usage"><code>(blastpgm fasta-file)</code></div><div class="doc"><pre class="plaintext">Determine blast program to use based on sequence alphabet of FASTA-FILE.
If amino acid alphabet, use tblastn, if nucleotide use blastn.  Returns
the function for one of these.</pre></div></div><div class="public anchor" id="var-canda"><h3>canda</h3><div class="usage"><code>(canda seqin cands-file canda-ofile)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-candf"><h3>candf</h3><div class="usage"><code>(candf in out &amp; k__1538__auto__)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-cands"><h3>cands</h3><div class="usage"><code>(cands in candf-output &amp; k__1538__auto__)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-cd-hit-est"><h3>cd-hit-est</h3><div class="usage"><code>(cd-hit-est input outfile &amp; {:keys [c n], :or {c 0.9, n 8}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-cmalign"><h3>cmalign</h3><div class="usage"><code>(cmalign cm seqfile outfile &amp; {:keys [opts par], :or {par 3}})</code></div><div class="doc"><pre class="plaintext">Run an automated parallelized cmalign on originating covariant
model CM (presumably obtained by some prior cmbuild/calibrate) and
fasta file SEQFILE.  Place new output sto in OUTFILE.  Return
outfile.  OPTS is any set of k/v pairs (as a vector) that are legal
for cmalign.

This is mostly (only?) used when building the next version of an
input sto that has gone through a complete cmbuild through cmsearch
through FFP and/or Scribl process yielding a new set of candidate
targets that have been placed in seqfile.
</pre></div></div><div class="public anchor" id="var-cmbuild"><h3>cmbuild</h3><div class="usage"><code>(cmbuild stofile &amp; {:keys [opts]})</code></div><div class="doc"><pre class="plaintext">Construct and issue an Infernal cmbuild command on alignment (sto!)
file stofile.  OPTS is a set of optional arguments given as either
a space separated string or a vector of strings.  Defaults to no
options (the -F option is automatically incorporated for the output
cmfile).  The output cmfile has a canonical name based on stofile -
&quot;.cm&quot; is appended as file type.

Ex: L20-auto-1.sto --&gt; L20-auto-1.sto.cm

Returns the cmfile spec as result.
</pre></div></div><div class="public anchor" id="var-cmcalibrate"><h3>cmcalibrate</h3><div class="usage"><code>(cmcalibrate cmfile &amp; {:keys [par opts], :or {par 3, opts (if (infernal-2+?) [&quot;--cpu&quot; &quot;4&quot;] [&quot;--mpi&quot;])}})</code></div><div class="doc"><pre class="plaintext">Construct and issue an Infernal cmcalibrate command on previously
built cm in cmfile.

PAR controls the number of MPI processes, _not_ the number of
threads.  Threads are only available on version 1.1+.  On thread
versions, the default is still to use mpi with an i/o thread (--cpu
1).

OPTS contains any other options that may be desired as either a
space separated string or as a vector of strings where all
parameters and their values are separate elements.

Returns the calibrated cmfile spec (same as input) as result.
</pre></div></div><div class="public anchor" id="var-cmf-post-process-combine"><h3>cmf-post-process-combine</h3><div class="usage"><code>(cmf-post-process-combine start-fna sto-glob)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-cmf-post-process-filter"><h3>cmf-post-process-filter</h3><div class="usage"><code>(cmf-post-process-filter insto &amp; {:keys [threshold], :or {threshold 10}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-cmfinder"><h3>cmfinder</h3><div class="usage"><code>(cmfinder seqin canda-file motif-out cm-out &amp; {:keys [initial-cm candf-output stdout]})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-cmfinder*"><h3>cmfinder*</h3><div class="usage"><code>(cmfinder* in &amp; {:keys [blastpgm blastdb initial-cm strand word-size max-cand-seq min-len-cand max-len-cand max-out-motifs expected-motif-freq min-stem-loops max-stem-loops del-files], :or {min-stem-loops 1, max-cand-seq 40, min-len-cand 30, del-files true, word-size 8, max-len-cand 100, strand :plus, max-out-motifs 4, max-stem-loops 2, expected-motif-freq 0.7}})</code></div><div class="doc"><pre class="plaintext">Perform a full run of cmfinder sub pipeline.  Basically:

(-&gt; in-fna (candf options)
    (#(if blastpgm) (blast options))
    (cands options)
    (map #(-&gt; % (canda in-fna) (cmfinder in-fna))))

Where options are the given set of optional keys and values (with
noted defaults).
</pre></div></div><div class="public anchor" id="var-cmsearch"><h3>cmsearch</h3><div class="usage"><code>(cmsearch cmfile fna-file outfile &amp; {:keys [par eval opts], :or {par 3, eval 1.0, opts (if (infernal-2+?) [&quot;--mid&quot; &quot;--cpu&quot; &quot;4&quot;] [&quot;--mpi&quot;])}})</code></div><div class="doc"><pre class="plaintext">Construct and issue an Infernal cmsearch command on previously
built/calibrated cm in cmfile against the seqs of interest in
&apos;database&apos; fna-file (a fasta file of sequences of interest).

PAR controls the number of MPI processes, _not_ the number of
threads.  Threads are only available on version 1.1+.  On thread
versions, the default is still to use mpi with an i/o thread (--cpu
1).

OPTS contains any other options that may be desired as either a
space separated string or as a vector of strings where all
parameters and their values are separate elements.

Output is placed in OUTFILE.  This outfile spec is returned as
result.
</pre></div></div><div class="public anchor" id="var-correct-dir-stos"><h3>correct-dir-stos</h3><div class="usage"><code>(correct-dir-stos stodir &amp; {:keys [dirdir], :or {dirdir false}})</code></div><div class="doc"><pre class="plaintext">
</pre></div></div><div class="public anchor" id="var-correct-sto-coordinates"><h3>correct-sto-coordinates</h3><div class="usage"><code>(correct-sto-coordinates stoin)</code><code>(correct-sto-coordinates sto1 sto2 &amp; stos)</code></div><div class="doc"><pre class="plaintext">
</pre></div></div><div class="public anchor" id="var-embl-sto-.3Enc-sto"><h3>embl-sto-&gt;nc-sto</h3><div class="usage"><code>(embl-sto-&gt;nc-sto stoin &amp; {:keys [blastout]})</code></div><div class="doc"><pre class="plaintext">Convert EMBL entry names to NCBI entry names while maintaining
alignment information.  STOIN is a sto file with entries named with
EMBL accesions.  Result is an output sto file with the alignments
of STOIN with NCBI NC accensions replacing the EMBL names.  Assumes
a nucleotide alphabet (no AA support yet).  Uses blast with 100%
identity match over full database and selects the hits with exact
size and content match.  If multiple candidates, selects randomly.
</pre></div></div><div class="public anchor" id="var-embl-to-nc"><h3>embl-to-nc</h3><div class="usage"><code>(embl-to-nc fin &amp; {:keys [blastout]})</code></div><div class="doc"><pre class="plaintext">Convert EMBL entry names to NCBI entry names while maintaining
sequence information.  FIN is either a sto or fna file with entries
named with EMBL accesions.  Result is a corresponding output sto or
fna file with the same sequence (and alignment if sto) with NCBI NC
accensions replacing the EMBL names.  Assumes a nucleotide
alphabet (no AA support yet).  Uses blast with 100% identity match
over full database and selects the hits with exact size and content
match.  If multiple candidates, selects randomly.
</pre></div></div><div class="public anchor" id="var-entry-aln-sqs"><h3>entry-aln-sqs</h3><div class="usage"><code>(entry-aln-sqs id-info idseq)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-entry-file-.3Eblastdb"><h3>entry-file-&gt;blastdb</h3><div class="usage"><code>(entry-file-&gt;blastdb efile &amp; {:keys [out ids], :or {out nil, ids nil}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-entry-file-difference"><h3>entry-file-difference</h3><div class="usage"><code>(entry-file-difference full f1 f2)</code><code>(entry-file-difference full f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Set difference the entry sets in the files f1 and f2 or f1, f2 and
remaining fs.  Full is true or false.  If true, entries must fully
match (name, start, end and strand) to be included in difference.
If false, only the names are used to match.  Returns the resulting
set of entries sans sequences.  See entry-file-difference-with-seqs
to include sequences as well.  File types may be fasta (fna,
hitfna), sto, aln, and gaisr csv.
</pre></div></div><div class="public anchor" id="var-entry-file-difference-with-seqs"><h3>entry-file-difference-with-seqs</h3><div class="usage"><code>(entry-file-difference-with-seqs f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Like entry-file-difference, but returns corresponding sqs for the
result entry set as well.  Uses the full entry (full is true for
entry-file-difference).  Difference the entries in files f1 f2, and
if non empty those in fs as well _in that order_.  The result is
thus the total &apos;subtraction&apos; of all elements from all fi, i /= 1,
from f1.  Take the result set and create the set of pairs [entry
sq] for each entry in the set.  Returns the resulting sequence of
pairs.
</pre></div></div><div class="public anchor" id="var-entry-file-intersect"><h3>entry-file-intersect</h3><div class="usage"><code>(entry-file-intersect full f1 f2)</code><code>(entry-file-intersect full f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Intersect the entry sets in the files f1 and f2 or f1, f2 and
remaining fs.  Full is true or false.  If true, entries must fully
match (name, start, end and strand) to be included in intersection.
If false, only the names are used to match.  Returns the resulting
set of entries sans sequences.  See entry-file-intersect-with-seqs
to include sequences as well.  File types may be fasta (fna,
hitfna), sto, aln, and gaisr csv.
</pre></div></div><div class="public anchor" id="var-entry-file-intersect-with-seqs"><h3>entry-file-intersect-with-seqs</h3><div class="usage"><code>(entry-file-intersect-with-seqs f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Like entry-file-intersect, but returns corresponding sqs for the
result entry set as well.  Uses the full entry (full is true for
entry-file-intersect).  Intersect the entries in files f1 f2, and
if non empty those in fs as well.  Take the result set and create
the set of pairs [entry sq] for each entry in the set.  Returns the
resulting sequence of pairs.
</pre></div></div><div class="public anchor" id="var-entry-file-setop"><h3>entry-file-setop</h3><h4 class="type">multimethod</h4><div class="usage"></div><div class="doc"><pre class="plaintext">Generic interface to entry file set operations with and without seqs
</pre></div></div><div class="public anchor" id="var-entry-file-union"><h3>entry-file-union</h3><div class="usage"><code>(entry-file-union full f1 f2)</code><code>(entry-file-union full f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Union the entry sets in the files f1 and f2 or f1, f2 and remaining
fs.  Full is true or false.  If true, entries must fully
match (name, start, end and strand) to be included in union.  If
false, only the names are used to match.  Returns the resulting set
of entries sans sequences.  See entry-file-union-with-seqs to
include sequences as well.  File types may be fasta (fna, hitfna),
sto, aln, and gaisr csv.
</pre></div></div><div class="public anchor" id="var-entry-file-union-with-seqs"><h3>entry-file-union-with-seqs</h3><div class="usage"><code>(entry-file-union-with-seqs f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Like entry-file-union, but returns corresponding sqs for the result
entry set as well.  Uses the full entry (full is true for
entry-file-union).  Union the entries in files f1 f2, and if non
empty those in fs as well.  Take the result set and create the set
of pairs [entry sq] for each entry in the set.  Returns the
resulting sequence of pairs.
</pre></div></div><div class="public anchor" id="var-entry-range-.3Efasta-file"><h3>entry-range-&gt;fasta-file</h3><div class="usage"><code>(entry-range-&gt;fasta-file entry range blastdb &amp; [filespec])</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-filter-base-hits-to-most-likely"><h3>filter-base-hits-to-most-likely</h3><div class="usage"><code>(filter-base-hits-to-most-likely blast-output-file)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-fuzzy-entry-file-difference"><h3>fuzzy-entry-file-difference</h3><div class="usage"><code>(fuzzy-entry-file-difference delta f1 f2)</code><code>(fuzzy-entry-file-difference delta f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Fuzzy difference of entry files.  By &apos;fuzzy&apos; we mean exact name
match of entries, exact strand match of entries, but start and end
within DELTA of f1 versions.</pre></div></div><div class="public anchor" id="var-fuzzy-entry-file-intersect"><h3>fuzzy-entry-file-intersect</h3><div class="usage"><code>(fuzzy-entry-file-intersect delta f1 f2)</code><code>(fuzzy-entry-file-intersect delta f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Fuzzy intersection of entry files.  By &apos;fuzzy&apos; we mean exact name
match of entries, exact strand match of entries, but start and end
within DELTA of f1 versions.</pre></div></div><div class="public anchor" id="var-fuzzy-entry-file-union"><h3>fuzzy-entry-file-union</h3><div class="usage"><code>(fuzzy-entry-file-union delta f1 f2)</code><code>(fuzzy-entry-file-union delta f1 f2 &amp; fs)</code></div><div class="doc"><pre class="plaintext">Fuzzy union of entry files.  By &apos;fuzzy&apos; we mean exact name
match of entries, exact strand match of entries, but start and end
within DELTA of f1 versions.</pre></div></div><div class="public anchor" id="var-genome-seq-dup.3F"><h3>genome-seq-dup?</h3><div class="usage"><code>(genome-seq-dup? coord1 coord2 &amp; {:keys [cutoff], :or {cutoff 0.8}})</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-blast-id-info-new"><h3>get-blast-id-info-new</h3><div class="usage"><code>(get-blast-id-info-new blast-output-file)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-cd-hit-infile"><h3>get-cd-hit-infile</h3><div class="usage"><code>(get-cd-hit-infile x)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-cmfsto-glob"><h3>get-cmfsto-glob</h3><div class="usage"><code>(get-cmfsto-glob motifstos)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-get-embl-blast-candidates"><h3>get-embl-blast-candidates</h3><div class="usage"><code>(get-embl-blast-candidates blast-output-file embl-seq-size-map)</code></div><div class="doc"><pre class="plaintext">Takes an ncbi blast csv output file result from blasting an embl
entry input and returns an embl-id to ncbi-ids map.  The csv has
fields:

  qseqid qstart qend evalue sseqid start send

The return map has entries: [embl-id [vec-of-best-ncbi-entries]].
</pre></div></div><div class="public anchor" id="var-get-entry-set"><h3>get-entry-set</h3><div class="usage"><code>(get-entry-set full f)</code></div><div class="doc"><pre class="plaintext">Helper function for &apos;entry file&apos; set operations.  Creates a set of
entries from the file F (fna, hitfna, sto, aln, and gaisr csv) with
name, start, stop and strand information if FULL is true or just
names if FULL is false.
</pre></div></div><div class="public anchor" id="var-get-tool-path"><h3>get-tool-path</h3><div class="usage"><code>(get-tool-path toolset-type)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-infernal-2.2B.3F"><h3>infernal-2+?</h3><div class="usage"><code>(infernal-2+?)</code></div><div class="doc"><pre class="plaintext">Determine whether the version of Infernal being called upon is post
1.0 (1.1+ basically).  We need this as later versions differ in
both their output format and the command arguments that they
use (in particular, --cpu, for thread usage control).
</pre></div></div><div class="public anchor" id="var-mk-infernal-cmd"><h3>mk-infernal-cmd</h3><div class="usage"><code>(mk-infernal-cmd cmd opts)</code></div><div class="doc"><pre class="plaintext">Build and check the canonical form of an Infernal command for us.
Checks that the command exists on the path, the version of Infernal
and cpu use (if later than 1.0.*), and folds in options.  cmd is
the base command as string.  opts is either a string of space
separated options (just like on cmdline) or a vector where all
pieces are separate elements.

Ensures command exists; returns the vector for command appropriate
for runx.
</pre></div></div><div class="public anchor" id="var-parse-blast"><h3>parse-blast</h3><div class="usage"><code>(parse-blast in out)</code></div><div class="doc"><pre class="plaintext">Parse a (NCBI) blastn output for use by cmfinder tools.  The blast run
specified output format with the options:

  -outfmt &apos;10 qseqid qstart qend evalue sseqid start send &lt;&amp; others&gt;&apos;

IN is the blast output filespec (path or file obj) and OUT is a filespec
(path or file obj) of where to place the parsed output.  This file
is then intended to be used as input to the CANDS program of
cmfinder that generates per motif-loop data files for input to
cmfinder.
</pre></div></div><div class="public anchor" id="var-process-blast-for-cmfinder"><h3>process-blast-for-cmfinder</h3><div class="usage"><code>(process-blast-for-cmfinder acc line)</code></div><div class="doc"><pre class="plaintext">A simple reducer for blastn output where the blast run specified
output format with the options:

-outfmt &apos;10 qseqid qstart qend evalue sseqid start send&apos;</pre></div></div><div class="public anchor" id="var-rank-em"><h3>rank-em</h3><div class="usage"><code>(rank-em summary-maps)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-refseq-entry.3F"><h3>refseq-entry?</h3><div class="usage"><code>(refseq-entry? entry-nm)</code></div><div class="doc"><pre class="plaintext">Returns whether entry-nm is a valid refseq entry for our
purposes (which bascially means it is in archea or bacteria...)
Returns nil if not, generalized true if yes.
</pre></div></div><div class="public anchor" id="var-summary-map"><h3>summary-map</h3><div class="usage"><code>(summary-map motif-sto-file)</code></div><div class="doc"><pre class="plaintext"></pre></div></div><div class="public anchor" id="var-tblastn"><h3>tblastn</h3><div class="usage"><code>(tblastn in &amp; kw-args)</code></div><div class="doc"><pre class="plaintext">NCBI tblastn (from the blast+ toolset).

IN is the input filespec in fasta format.

KEYWORD ARGS:
:out is the output filespec for direct blast output, default `in`.blast
:blastdb the filespec for the preformatted blast database to use
:word-size blast bp slice size (chunk size for alignment), default 8
:evalue (real) expectation (E) value threshold for saving hits, default 10
:fmt NCBI blastn out fmt string, defaults to:
     10 qseqid qstart qend evalue sseqid sstart send
:fn post processing fn of at least 1 arg for blast output, eg parse-blast
:fnargs any other args for fn, e.g., output filespec</pre></div></div></div></body></html>